### DISCUSSION TITLE
Requirement for read() command in QCoDeS and reasons

### DISCUSSION THREAD

**Discussion by y3yang** [microsoft/Qcodes#4158] [Ideas]
Hi, although pyvisa provides write(), query() and read(), I can only find two of them in QCoDeS. Unfortunately, there is a situation that I need read() in QCoDeS.

I'm trying to write a driver for keithley 6221. And I found that I will get two line feeds ('\n' as terminator) when I ask keithley 6221 to get the parameter of keithley 2182A, So I need read() command to eliminate the effect of the second line feed. Otherwise, it will affect other commands.

Therefore, I hope to know if there will be read() command in the future. Or could you please give me some advice about how to add a read() command in "InstrumentChannel"? 

Details:
We have Keithley 6221 and Keithely 2182A to do differential conductance measurement. The way to control Keithley 2182A is sending command to Keithley 6221 by GPIB and Keithley 6221 will send command to Keithley 2182 by RS-232. Then Keithely 2182 will send message back to Keithley 6221. Finally, I need to ask Keithley 6221 to get the message. This message contains two line feeds.

![image](https://user-images.githubusercontent.com/79177612/167254790-5cf631d3-aea3-49f8-8b25-5af8dfd7ebb2.png)

**Comment by jenshnielsen** [comment]
Hi @y3yang 

I think there are a few different ways to handle this. 

All QCoDeS visa instruments have exposes the pyvisa visa handle directly
so you are free to implement your own read command in the instrument class.  

in the 6220  you could add something like


```
def read(self) -> str:
     return self.visa_handle.read()
```

E.g. similar to how the write_raw and ask_raw are implemented in the regular visa driver

If the instrument always returns a double newline on read it may also make sense to set the read terminator to that?
